File: rust/src/medium/robot_simulator.rs
#[derive(PartialEq, Eq, Debug)]
pub enum Direction {
    North,
    East,
    South,
    West,
}

pub struct Robot {
    x: i32,
    y: i32,
    direction: Direction,
}

impl Robot {
    pub fn new(x: i32, y: i32, direction: Direction) -> Self {
        Robot { x, y, direction }
    }

    pub fn turn_right(self) -> Self {
        use Direction::*;
        let new_direction = match self.direction {
            North => East,
            East => South,
            South => West,
            West => North,
        };

        Robot {
            direction: new_direction,
            ..self
        }
    }

    pub fn turn_left(self) -> Self {
        use Direction::*;
        let new_direction = match self.direction {
            North => West,
            West => South,
            South => East,
            East => North,
        };

        Robot {
            direction: new_direction,
            ..self
        }
    }

    pub fn advance(self) -> Self {
        use Direction::*;
        let new_coords = match self.direction {
            North => (self.x, self.y + 1),
            West => (self.x - 1, self.y),
            South => (self.x, self.y - 1),
            East => (self.x + 1, self.y),
        };

        Robot {
            x: new_coords.0,
            y: new_coords.1,
            ..self
        }
    }

    pub fn instructions(self, instructions: &str) -> Robot {
        instructions.chars().fold(self, |acc, char| match char {
            'R' => acc.turn_right(),
            'L' => acc.turn_left(),
            'A' => acc.advance(),
            _ => panic!("Action is unknown: {}", char),
        })
    }

    pub fn position(&self) -> (i32, i32) {
        (self.x, self.y)
    }

    pub fn direction(&self) -> &Direction {
        &self.direction
    }
}

#[test]
fn turning_right_from_south_points_the_robot_west() {
    let robot = Robot::new(0, 0, Direction::South).turn_right();

    assert_eq!(&Direction::West, robot.direction());
}

#[test]
fn follow_instructions_to_move_east_and_north() {
    let robot = Robot::new(8, 4, Direction::South).instructions("LAAARRRALLLL");

    assert_eq!((11, 5), robot.position());

    assert_eq!(&Direction::North, robot.direction());
}

File: elixir/lib/medium/robot_simulator.ex
defmodule RobotSimulator do
  @directions [:north, :east, :south, :west]

  defstruct [:direction, :position]

  def create(direction \\ :north, position \\ {0, 0})

  def create(direction, _) when direction not in @directions, do: {:error, "invalid direction"}

  def create(direction, {x, y}) when is_integer(x) and is_integer(y) do
    %RobotSimulator{direction: direction, position: {x, y}}
  end

  def create(_direction, _), do: {:error, "invalid position"}

  def simulate(robot, instructions) do
    instructions = String.codepoints(instructions)

    Enum.reduce_while(instructions, robot, fn instruction, robot ->
      case instruction do
        "R" -> {:cont, turn_right(robot)}
        "L" -> {:cont, turn_left(robot)}
        "A" -> {:cont, advance(robot)}
        _ -> {:halt, {:error, "invalid instruction"}}
      end
    end)
  end

  defp turn_right(%{direction: :north} = robot), do: %{robot | direction: :east}
  defp turn_right(%{direction: :east} = robot), do: %{robot | direction: :south}
  defp turn_right(%{direction: :south} = robot), do: %{robot | direction: :west}
  defp turn_right(%{direction: :west} = robot), do: %{robot | direction: :north}

  defp turn_left(%{direction: :north} = robot), do: %{robot | direction: :west}
  defp turn_left(%{direction: :west} = robot), do: %{robot | direction: :south}
  defp turn_left(%{direction: :south} = robot), do: %{robot | direction: :east}
  defp turn_left(%{direction: :east} = robot), do: %{robot | direction: :north}

  defp advance(%{direction: direction, position: {x, y}} = robot) do
    case direction do
      :north -> %{robot | position: {x, y + 1}}
      :east -> %{robot | position: {x + 1, y}}
      :south -> %{robot | position: {x, y - 1}}
      :west -> %{robot | position: {x - 1, y}}
    end
  end

  def direction(%RobotSimulator{direction: direction}), do: direction
  def position(%RobotSimulator{position: position}), do: position
end
